File: striver/S2L2/quick.java
package S2L2;

import java.util.Scanner;

public class quick {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int[] arr = new int[n];
        for(int i=0;i<n;i++)
            arr[i] = sc.nextInt();
        qs(arr,0,n-1);
        for(int i=0;i<n;i++)
            System.out.print(arr[i]+" ");
    }
    public static void qs(int[] arr,int l,int r){
        if(l<r){
        int pi = fixedpoint(arr,l,r);
        qs(arr,l,pi-1);
        qs(arr,pi+1,r);
        } 
    }
    // public static int fixedpoint(int[] arr,int l,int r){
    //     int[] narr = new int[arr.length];
    //     int temp = arr[l];
    //     int lp = l,rp = r;
    //     for(int i=l+1;i<=r;i++){
    //         if(arr[i]<=temp){
    //             narr[lp]=arr[i];
    //             lp++;
    //         } else {
    //             narr[rp]=arr[i];
    //             rp--;
    //         }
    //     }
    //     narr[lp]=temp;
    //     for(int i=l;i<=r;i++){
    //         arr[i]=narr[i];
    //     }
    //     return lp;
    // }
    public static int fixedpoint(int[] arr,int l,int r){
        int temp = arr[l];
        int lp = l+1;
        int rp = r;
        while(lp<rp){
            while(lp<rp && temp>=arr[lp])lp++;
            while(temp < arr[rp])rp--;
            if(lp<rp){
                int t = arr[lp];
                arr[lp] = arr[rp];
                arr[rp] =t;
            }
        }
        arr[l] = arr[rp];
        arr[rp] = temp;
        return rp;
    }
}

File: striver/S2L2/quick.py
size = int(input())
arr = [0]*size
for i in range(0,size):
    arr[i]=int(input())
# def fixedpoint(arr,l,r):
#     narr = [0]*len(arr)
#     temp = arr[l]
#     lp = l
#     rp = r
#     for i in range(l+1,r+1):
#         if temp >= arr[i]:
#             narr[lp] = arr[i]
#             lp+=1
#         else:
#             narr[rp] = arr[i]
#             rp-=1
#     narr[lp] = temp
#     for i in range(l,r+1):
#         arr[i]=narr[i]
#     return lp
def fixedpoint(arr,l,r):
    temp = arr[l]
    lp = l+1
    rp = r
    while lp<rp :
            while lp<rp and temp>=arr[lp]:
                lp+=1
            while temp < arr[rp] :
                rp-=1
            if lp<rp:
                t = arr[lp]
                arr[lp] = arr[rp]
                arr[rp] =t
    arr[l] = arr[rp]
    arr[rp] = temp
    return rp       
def quick_sort(arr,l,r):
    if l<r:
        pi = fixedpoint(arr,l,r)
        quick_sort(arr,l,pi-1)
        quick_sort(arr,pi+1,r)
        
quick_sort(arr,0,size-1)
for i in range(0,size):
        print(arr[i],end=" ")